### PR TITLE
[refactor](transaction) extract the majority rules of transaction

### PR DESCRIPTION
**Pull request by yujun777**
## Proposed changes

1. extract the majority rule of transaction;
2. fix new join replicas may treate as succefully even if they don't  commit the transaction;

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by yujun777** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/transaction/PublishVersionDaemon.java [hunk 1]
```diff
@@ -63,13 +58,33 @@ protected void runAfterCatalogReady() {
...
+        Map<Long, Set<Long>> publishErrorReplicas = Maps.newHashMap();
+        for (PublishVersionTask task : transactionState.getPublishVersionTasks().values()) {
+            long backendId = task.getBackendId();
+            if (!task.isFinished()) {
+                continue;
+            }
+            Set<Long> errorReplicas = Sets.newHashSet();
+            for (Long tabletId : task.getErrorTablets()) {
+                if (tabletInvertedIndex.getTabletMeta(tabletId) == null) {
+                    continue;
```
Comment (dataroaring) [inline fe/fe-core/src/main/java/org/apache/doris/transaction/PublishVersionDaemon.java:72]: We should print a log here, it means a balance or drop operation happens.

File: fe/fe-core/src/main/java/org/apache/doris/transaction/PublishVersionDaemon.java [hunk 2]
```diff
@@ -63,13 +58,33 @@ protected void runAfterCatalogReady() {
...
+                }
+                Replica replica = tabletInvertedIndex.getReplica(tabletId, backendId);
+                if (replica != null) {
+                    errorReplicas.add(replica.getId());
+                }
```
Comment (dataroaring) [inline fe/fe-core/src/main/java/org/apache/doris/transaction/PublishVersionDaemon.java:77]: else {
    // We should print a log here, it means a balance operation happens.

}

**Review by dataroaring — DISMISSED** [review]
LGTM

File: fe/fe-core/src/main/java/org/apache/doris/transaction/PublishVersionDaemon.java (hunk 1 above)
Comment (yujun777) [reply fe/fe-core/src/main/java/org/apache/doris/transaction/PublishVersionDaemon.java:74]: done print log

File: fe/fe-core/src/main/java/org/apache/doris/transaction/PublishVersionDaemon.java (hunk 2 above)
Comment (yujun777) [reply fe/fe-core/src/main/java/org/apache/doris/transaction/PublishVersionDaemon.java:83]: done print log

**Comment by yujun777** [comment]
run buildall

**Comment by yujun777** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 44.94 seconds
 stream load tsv:          506 seconds loaded 74807831229 Bytes, about 140 MB/s
 stream load json:         19 seconds loaded 2358488459 Bytes, about 118 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.4 seconds inserted 10000000 Rows, about 340K ops/s
 storage size: 17167545296 Bytes

**Comment by yujun777** [comment]
#24273  has fix publish txn bug.

**Pull request closed without merging**